versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/CommitRetry.java

```diff
@@ -60,6 +61,10 @@ static <T> T commitRetry(Persist persist, CommitAttempt<T> attempt, TryLoopState
           throw new RetryTimeoutException(i, tls.currentNanos() - t0);
         }
         retryState = e.retryState();
+      } catch (UnknownOperationResultException e) {
+        if (!tls.retry(t1)) {
```

Comment:
I'm not sure about retrying the whole commit in this case. Suppose the timeout occurs, but the change is still persisted by the database... Would it make sense to try reading the reference HEAD to check whether it's advanced?

In case of a persisted write with a timeout, if we retry building the commit, it will conflict with its own previous (persisted and readable) state and result in a cryptic conflict error at the user level.. WDYT?

Reply:
I think we should rebuild the commit object only if the reference HEAD changes between retries. WDYT?

Reply:
> Would it make sense to try reading the reference HEAD to check whether it's advanced?

That's what `BaseCommitHelper.committingOperation()` does, no?

> In case of a persisted write with a timeout, if we retry building the commit, it will conflict with its own previous (persisted and readable) state and result in a cryptic conflict error at the user level.

How should it conflict, if the ref HEAD didn't change?

Reply:
Updated the PR (added some more test cases) - basically the existing 3 existing cases, which "timeout" without changing the persisted state, but with a "timeout" after changing the persisted state.

* For `storeObjs` (content objects) - that's a "no brainer"
* For storing the commit itself it result(ed) in a "hash collision" - but that could be worked around - with a related change for transplant operations ("fast forward transplants") - not breaking tho - so should be fine.
* For bumping the reference it's trickier - the workaround in this change works often - but won't work for concurrent commits to the same same reference that run into timeouts. I don't see a way to guarantee the "right" outcome without adding a ton of complexity - which might not even work. I think this falls into the "_Adventures in distributed computing_" category... - probably okay to leave it as it is.